pkg/v1/tkg/web/src/app/views/landing/wizard/shared/components/steps/metadata-step/metadata-step.component.ts

```diff
@@ -37,154 +29,39 @@ export class MetadataStepComponent extends StepFormDirective implements OnInit {
-        const control = this.formGroup.get(dependentFieldName);
-        this.registerOnValueChange(fieldName, (data) => {
-            if (data !== '') {
-                if (!control.hasValidator(Validators.required)) {
-                    control.addValidators(Validators.required);
-                    control.markAsPending(); // validation will not be triggered until the field is touched.
-                    control.setErrors({required: true});
-                }
-            } else {
-                control.removeValidators(Validators.required);
+        this.tkgLabelsConfig = {
+            label: {
+                title: this.htmlFieldLabels['clusterLabels'],
+                tooltipText: `Optionally specify labels for the ${this.clusterTypeDescriptor} cluster.`,
+            },
+            forms: {
+                parent: this.formGroup,
+                control: this.formGroup.get('clusterLabels') as FormArray,
+            },
+            fields: {
+                clusterType: this.clusterTypeDescriptorTitleCase,
+                fieldMapping: MetadataStepMapping.fieldMappings.find((m) => m.name === MetadataField.CLUSTER_LABELS),
             }
-            this.validateAllLabels(); // all the same label keys can show error message.
-        });
-    }
-
-    validateAllLabels () {
-        // The setTimeout wrapper ensures that validation logic will run after a new label field is added.
-        setTimeout(_ => {
-            for (const [labelKey, labelVal] of this.labels) {
-                const key = this.formGroup.get(labelKey);
-                const val = this.formGroup.get(labelVal);
-                if (key) {
-                    if (this.savedKeySet.has(labelKey)) {
-                        key.markAsTouched();
-                    }
-                    key.updateValueAndValidity();
-                }
-                if (val) {
-                    val.updateValueAndValidity();
-                }
-            }
-        });
-    }
-
-    deleteLabel(key: string) {
-        this.formGroup.removeControl(key);
-        this.formGroup.removeControl(this.labels.get(key));
-        this.labels.delete(key);
-        this.keySet.delete(key);
+        };
     }
 
     dynamicDescription(): string {
         const clusterLocation = this.getFieldValue(MetadataField.CLUSTER_LOCATION, true);
-        return clusterLocation ? 'Location: ' + clusterLocation : 'Specify metadata for the ' + this.clusterTypeDescriptor + ' cluster';
+        return clusterLocation
+            ? 'Location: ' + clusterLocation
+            : 'Specify metadata for the ' + this.clusterTypeDescriptor + ' cluster';
     }
 
     protected storeUserData() {
         this.storeUserDataFromMapping(this.supplyStepMapping());
         this.storeDefaultDisplayOrder(this.supplyStepMapping());
     }
+
+    private supplyStepMapping(): StepMapping {
+        return this.stepMapping ?? MetadataStepMapping;
```

Comment:
I'm not sure it makes sense to return the default MetadataStepMapping if this.stepMapping is null/undefined. Is this.stepMapping === null/undefined a legitimate state?

Reply:
refactored